### PR TITLE
bpo-47177: Replace `f_lasti` with `prev_instr`

### PR DESCRIPTION
**Pull request by brandtbucher**
A nice 2% improvement:

<details>

```
Slower (2):
- telco: 38.6 ms +- 2.7 ms -> 40.9 ms +- 2.6 ms: 1.06x slower
- regex_dna: 1.33 sec +- 0.01 sec -> 1.34 sec +- 0.01 sec: 1.01x slower

Faster (32):
- logging_silent: 640 ns +- 27 ns -> 582 ns +- 35 ns: 1.10x faster
- scimark_sor: 736 ms +- 16 ms -> 689 ms +- 13 ms: 1.07x faster
- pycparser: 7.52 sec +- 0.10 sec -> 7.12 sec +- 0.10 sec: 1.06x faster
- chameleon: 39.8 ms +- 2.7 ms -> 37.8 ms +- 2.4 ms: 1.05x faster
- html5lib: 417 ms +- 35 ms -> 397 ms +- 31 ms: 1.05x faster
- scimark_monte_carlo: 422 ms +- 13 ms -> 402 ms +- 11 ms: 1.05x faster
- pyflate: 2.65 sec +- 0.02 sec -> 2.53 sec +- 0.02 sec: 1.05x faster
- deltablue: 22.5 ms +- 1.3 ms -> 21.6 ms +- 1.4 ms: 1.05x faster
- unpickle_pure_python: 1.43 ms +- 0.08 ms -> 1.37 ms +- 0.07 ms: 1.04x faster
- spectral_norm: 628 ms +- 10 ms -> 604 ms +- 13 ms: 1.04x faster
- hexiom: 41.0 ms +- 3.1 ms -> 39.5 ms +- 2.4 ms: 1.04x faster
- raytrace: 1.84 sec +- 0.02 sec -> 1.78 sec +- 0.02 sec: 1.03x faster
- scimark_lu: 669 ms +- 20 ms -> 648 ms +- 20 ms: 1.03x faster
- xml_etree_generate: 474 ms +- 11 ms -> 459 ms +- 13 ms: 1.03x faster
- richards: 296 ms +- 12 ms -> 287 ms +- 14 ms: 1.03x faster
- nbody: 555 ms +- 17 ms -> 540 ms +- 15 ms: 1.03x faster
- pickle_dict: 170 us +- 8 us -> 165 us +- 7 us: 1.03x faster
- crypto_pyaes: 510 ms +- 12 ms -> 497 ms +- 13 ms: 1.03x faster
- scimark_fft: 1.99 sec +- 0.03 sec -> 1.95 sec +- 0.02 sec: 1.02x faster
- regex_compile: 819 ms +- 11 ms -> 800 ms +- 10 ms: 1.02x faster
- chaos: 423 ms +- 12 ms -> 414 ms +- 10 ms: 1.02x faster
- meteor_contest: 643 ms +- 13 ms -> 630 ms +- 12 ms: 1.02x faster
- pidigits: 1.18 sec +- 0.01 sec -> 1.16 sec +- 0.01 sec: 1.02x faster
- sympy_expand: 2.86 sec +- 0.02 sec -> 2.81 sec +- 0.02 sec: 1.02x faster
- nqueens: 498 ms +- 13 ms -> 490 ms +- 13 ms: 1.02x faster
- xml_etree_parse: 953 ms +- 20 ms -> 938 ms +- 16 ms: 1.02x faster
- 2to3: 1.60 sec +- 0.01 sec -> 1.58 sec +- 0.01 sec: 1.02x faster
- xml_etree_process: 331 ms +- 12 ms -> 326 ms +- 11 ms: 1.02x faster
- sympy_str: 1.71 sec +- 0.02 sec -> 1.69 sec +- 0.02 sec: 1.01x faster
- fannkuch: 2.42 sec +- 0.02 sec -> 2.39 sec +- 0.02 sec: 1.01x faster
- dulwich_log: 381 ms +- 11 ms -> 377 ms +- 11 ms: 1.01x faster
- xml_etree_iterparse: 624 ms +- 14 ms -> 618 ms +- 13 ms: 1.01x faster

Benchmark hidden because not significant (28): django_template, float, go, json, json_dumps, json_loads, logging_format, logging_simple, mako, pathlib, pickle, pickle_list, pickle_pure_python, python_startup, python_startup_no_site, regex_effbot, regex_v8, scimark_sparse_mat_mult, sqlalchemy_declarative, sqlalchemy_imperative, sqlite_synth, sympy_integrate, sympy_sum, thrift, tornado_http, unpack_sequence, unpickle, unpickle_list

Geometric mean: 1.02x faster
```

</details>

https://bugs.python.org/issue47177

### CODE REVIEW

File: Include/internal/pycore_frame.h
Comment (iritkatriel) [inline Include/internal/pycore_frame.h:63]: The i in lasti referred to 'instruction', but now it's the previous word which could be in the cache of that instruction, so should this macro be named something else?

File: Modules/_tracemalloc.c
Comment (markshannon) [inline Modules/_tracemalloc.c]: Maybe add `_PyInterpreterFrame_GetLine()`?

File: Objects/frameobject.c
Comment (markshannon) [inline Objects/frameobject.c]: Could use `_PyInterpreterFrame_GetLine()` here as well

File: Objects/frameobject.c
Comment (markshannon) [inline Objects/frameobject.c]: Off by one? The test was `i < frame->f_lasti` the new test is equivalent to `i < frame->f_lasti+1`, I think.

File: Objects/genobject.c
Comment (markshannon) [inline Objects/genobject.c]: I think this can be replaced with `gen->gi_frame_state  == FRAME_CREATED`.

File: Objects/genobject.c
Comment (markshannon) [inline Objects/genobject.c]: Slightly off-topic, but I wonder why we are going to the expense of computing the line here, rather than just storing the instruction offset?

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: Here's something to consider. 

Rather that storing `next_instr`, we could store `last_instr` (strictly, `next_instr_less_one`) as it would break the dependency of the store on the addition.
`frame->next_instr_less_one = next_instr; next_instr++`
The compiler can then merge the `next_instr++` with any subsequent `JUMPBY(INLINE_CACHE_ENTRIES_...)`

It would also simplify the changes as you wouldn't need to change all the offsets by one.

Thoughts?

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: There are a few `frame->next_instr += X` here and below.

Provided `frame->next_instr` is up to date with `next_instr`, this would be more efficient as
```
JUMPBY(X); frame->next_instr = next_instr
``` 
It is possible the compiler can work that out for itself, but I think the `JUMPBY` version is a bit clearer. Maybe add a `SAVE_IP()` macro so this becomes `JUMPBY(X); SAVE_IP();`

**Review by markshannon — CHANGES_REQUESTED** [review]
Looks very promising.

I think there is an off by one error, and I have a few suggestions and questions.

**Comment by markshannon** [comment]
[My performance test](https://gist.github.com/markshannon/62f3d9f37946085066f1fc38140c1895) shows the same 2% but with less variation, and no 6% slowdown for `telco`.

File: Include/internal/pycore_frame.h
Comment (brandtbucher) [reply Include/internal/pycore_frame.h:63]: I mainly named it `LASTI` to make its meaning clear for anyone familiar with the prior code. Also, the `f_lasti` member of Python-layer frame objects has the same value (and we're not renaming that), so to me it makes sense to keep the names in sync.

I'll add a clarifying comment, though.

File: Python/frame.c
Comment (arhadthedev) [inline Python/frame.c]: No newline at end of file.

File: Objects/genobject.c
Comment (brandtbucher) [reply Objects/genobject.c]: Can it? This seems like it also handles the case where `f_lasti == 0`, meaning that the frame has actually started.

I *think* we could end up in a case like that if a generator is closed or something while handling an eval-breaker in its initial `RESUME`. But it looks like that case might actually be handled by the next branch below?

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: Interesting. I'll try that out in another branch.

File: Objects/genobject.c
Comment (markshannon) [reply Objects/genobject.c]: The 0th instruction is `RETURN_GENERATOR`, so the frame hasn't started until (at least) the 1st (index 1) instruction.

Oh the joys of counting from zero, in English.

```
def gen(): yield
```
```
  1           0 RETURN_GENERATOR
              2 POP_TOP
              4 RESUME                   0
              6 LOAD_CONST               0 (None)
              8 YIELD_VALUE
             10 RESUME                   1
             12 POP_TOP
             14 LOAD_CONST               0 (None)
             16 RETURN_VALUE
```

Comment (brandtbucher) [reply Objects/genobject.c]: Ah, forgot about `RETURN_GENERATOR`. Thanks for clarifying.

File: Include/internal/pycore_frame.h
Comment (sweeneyde) [inline Include/internal/pycore_frame.h]:
```suggestion
// For example, it may be the offset of an inline CACHE entry, an instruction we
```

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: https://github.com/brandtbucher/cpython/compare/lasti...brandtbucher:lasti-alt

Comment (brandtbucher) [reply Python/ceval.c]: The performance numbers show no difference between that branch and this one. The other branch does result in a net reduction of "adjust by 1" moves, so I think I slightly prefer it.

Thoughts?

Comment (markshannon) [reply Python/ceval.c]: I prefer the other version. It should be quicker, and feels less intrusive. I'm not surprised that there is no measurable difference in performance, I would expect it to be very slight.

**Comment by markshannon** [comment]
Did you get the version with `prev_instr` ready for review/merging?

**Comment by brandtbucher** [comment]
It's ready, just slipped off my radar. Would you prefer a separate PR, or just to bring those commits over to this branch?

**Comment by markshannon** [comment]
Whatever is easier. I don't mind

**Comment by brandtbucher** [comment]
I brought the commits over here.

**Comment by markshannon** [comment]
Looks good. 
This is quite an intrusive change, so I'm running the buildbots before merging.

**Comment by brandtbucher** [comment]
Looking into it.

**Comment by brandtbucher** [comment]
https://bugs.python.org/issue47250

**Comment by markshannon** [comment]
Feel free to merge once you are confident that all the failures are unrelated. I think they are unrelated, but I've not checked them individually.

**Comment by brandtbucher** [comment]
The refleaks disappear when I apply the fix in #32403. So not my fault. 🙂

**Comment by vstinner** [comment]
The coverage project uses the PyFrameObject.f_frame.f_lasti field and so is broken by this PR. I wrote https://github.com/nedbat/coveragepy/pull/1331 to read the Python attribute, but the coverage maintainer has worries about worse performance. Maybe adding a new PyFrame_GetLastInstr() getter function would help.

**Comment by brandtbucher** [comment]
The PR adds a `_PyInterpreterFrame_LASTI` macro. Does that help?

**Comment by markshannon** [comment]
https://github.com/python/cpython/pull/32413

**Pull request merged**
